hack/verify-staging-imports.sh

```diff
@@ -23,13 +23,38 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 
 kube::golang::setup_env
 
+function print_forbidden_imports () {
+    local PACKAGE="$1"
+    shift
+    local RE=""
+    local SEP=""
+    for CLAUSE in "$@"; do
+        RE+="${SEP}${CLAUSE}"
+        SEP='\|'
+    done
+    local FORBIDDEN=$(
+        go list -f $'{{with $package := .ImportPath}}{{range $.Imports}}{{$package}} imports {{.}}\n{{end}}{{end}}' ./vendor/k8s.io/${PACKAGE}/... |
+        sed 's|^k8s.io/kubernetes/vendor/||;s| k8s.io/kubernetes/vendor/| |' |
+        grep -v " k8s.io/${PACKAGE}" |
+        grep -e "imports \(${RE}\)"
+    )
+    if [ -n "${FORBIDDEN}" ]; then
+        echo "${PACKAGE} has a forbidden dependency:"
+		echo
+		echo "${FORBIDDEN}" | sed 's/^/  /'
+		echo
+		return 1
+    fi
+    return 0
+}
 
-for dep in $(ls -1 ${KUBE_ROOT}/staging/src/k8s.io/); do
-	if go list -f {{.Deps}} ./vendor/k8s.io/${dep}/... | tr " " '\n' | grep k8s.io/kubernetes | grep -v 'k8s.io/kubernetes/vendor' | LC_ALL=C sort -u | grep -e "."; then
-		echo "${dep} has a cyclical dependency"
-		exit 1
-	fi
-done
+RC=0
+print_forbidden_imports apimachinery k8s.io/ || RC=1
```

Comment:
If at any point you expect `set -o errexit` to alter the flow of commands inside of `print_forbidden_imports` -- like, a failure somewhere in there you would expect to exit early ... in this construct, it would not. `function_call || something` will turn off `errexit` inside the function and there is no way to turn it back on.